Add EffVer to release guide and link to checklist template (#839)
napari/docs: docs/developers/coredev/release.md

# Release guide

This guide documents `napari`'s release process as of 0.6.4.
Most required tools mentioned here are in https://github.com/napari/napari-release-tools.

:::{important}

> The first step in managing a release is to create a new Issue in napari/napari with the [`release` template](https://github.com/napari/napari/tree/main/.github/ISSUE_TEMPLATE/release.md).

Reviewer: wow cool! 🤩

# Release guide

This guide documents `napari`'s release process as of 0.6.4.
Most required tools mentioned here are in https://github.com/napari/napari-release-tools.

## EffVer: (Intended) Effort Versioning

As of 0.6.5, napari has officially adopted [EffVer](https://effver.org) for versioning, from the previous semantic versioning.
While in pre-1.0 state, the current convention is `0.MACRO.MESO` and once we release v1.0.0 will henceforth be in `MACRO.MESO.MICRO` versioning.
This is similar to previous Semantic Versioning schema of `MAJOR.MINOR.BUGFIX`, but allows features and bugfixes to live harmoniously in any version.

> Instead, the version number implies to users the amount of effort required to adopt the newest version from signficant effort (`MACRO`) to no expected effort (`MICRO`).

Reviewer 2: ```suggestion
Instead, the version number implies to users the amount of effort required to adopt the newest version from significant effort (`MACRO`) to no expected effort (`MICRO`).
```

Our current policy permits deprecations in `MACRO` and `MESO` releases only.

## Release Checklist

### Early management

- [ ] Make a new Zulip thread in the [release channel](https://napari.zulipchat.com/#narrow/stream/215289-release)
- [ ] Manage Github release Milestone
  - [ ] Set rc0 as the first due date of the milestone at least 3 days before. Given a typical release cycle, a week prior is recommended.
    - *Note the Milestone date is on [Line Island Time (LINT)](https://www.timeanddate.com/time/zone/@4030926) where the village of napari is located.*
  - [ ] Add the Zulip release thread to the milestone description

### Triage the Milestone

- [ ] Triage the milestone throughout the release cycle.
  - [ ] Add PRs of interest to highlights. These help the manager keep track of important PRs that often require extra team attention.

>   - [ ] Move PRs and issues that won't make it in the current release to the next milestone. Add a note to PRs (and issues) from community members describing the actions the release manager is taking and provide actionable comments.

Reviewer 2: And/or remove the milestone unless it's clear it will be merged soon

### Create Pre-release

*You may wish to make an alpha release at any point in the release cycle. However, you must make at least an empty release file to attach to the tag and merge it to `napari/docs`, otherwise the release will fail.*

- [ ] Communicate in Zulip release thread that pre-release is approaching.
- [ ] Create a draft of release notes and make PR to `napari/docs`.
  - [ ] Merge release notes to `napari/docs`

> - [ ] Tag pre-release with release notes file and push to `napari/napari`

Reviewer 2: This needs some explanation and/or command!

- [ ] Announce release candidate on Zulip in Release (and General, for meso or larger releases) and [forum.image.sc](https://forum.image.sc/announcements)
- [ ] Check for proper deployment
  - [ ] PyPI

>   - [ ] napari.org

Reviewer 2: how long does it take typically?

### Stable Release Prep

> 

Author: I should add in here about cutting any relevant releases to other repos, like napari-sphinx-theme, napari-plugin-manager, etc

Reviewer 2: Are there repos where we specifically always release in sync?

- [ ] Triage remaining PRs and Issues. At this stage, bug fixes and remaining (testable) features are prioritized.
- [ ] Make new PR to `napari/docs` with changes to release notes.
- [ ] Ensure [`conda-recipe/recipe.yaml`](https://github.com/napari/packaging/blob/main/conda-recipe/recipe.yaml) in `napari/packaging` is up-to-date (e.g. `run` dependencies match `pyproject.toml` requirements).

> - [ ] Ensure that `contraints` files are up to date. Usually initiated by `@napari-bot` within a day of changes to dependencies and otherwise regularly scheduled.

Reviewer 2: ```suggestion
- [ ] Ensure that `constraints` files are up to date. Usually initiated by `@napari-bot` within a day of changes to dependencies and otherwise regularly scheduled.
```

### Create Stable Release

- [ ] Do a final build, push, and merge of release notes to `napari/docs`.
- [ ] Ensure `napari/docs` successfully deploys
- [ ] Checkout `napari/napari:main` and tag with release notes file.
- [ ] Push created tag to `napari/napari`. This triggers the deployment actions.
- [ ] Check that deployment is successful
  - [ ] `napari/docs` and `napari/napari.github.io`
  - [ ] PyPI
  - [ ] Conda-forge [`napari-feedstock`](https://github.com/conda-forge/napari-feedstock). *Note: Conda-forge bots will automatically make a PR, usually within 12 hours.*

> - [ ] Update symlink in `napari/napari.github.io` using the [action](https://github.com/napari/napari.github.io/actions/workflows/symlink-stable.yml).

Reviewer 2: ```suggestion
- [ ] Update symlink in `napari/napari.github.io` by manually triggering the [action](https://github.com/napari/napari.github.io/actions/workflows/symlink-stable.yml).
```

- Create a zulip thread in [the release channel](https://napari.zulipchat.com/#narrow/stream/215289-release)
  letting people know the release candidate is coming and pointing out PRs that would be nice to merge before release.
- Look through currently open PRs and get a sense of what would be good to merge before the first release candidate. Set milestones appropriately;

> - Ensure `conda-recipe/meta.yaml` in `napari/packaging` is up-to-date (e.g. `run` dependencies match `pyproject.toml` requirements);

Reviewer: Why is this removed?

Author: mistake! I think because I didn't want to be redundant with the checklist I made, but I ripped that out to napari/napari template

Author: oh, actually its because I moved it down to the creating a prerlease section

- Add a header and highlights section to the [`additional notes`](https://github.com/napari/napari-release-tools/tree/main/additional_notes) folder for the given release.
  Use the [highlight label](https://github.com/napari/napari/pulls?q=sort%3Aupdated-desc+is%3Apr+is%3Aopen+label%3Ahighlight) for the relevant milestone to note which PRs to comment on.

>   To add images or GIFs (the only supported animated format), add the file as a comment to appropriate PR.  You should link to the public GitHub link for that file and not the private assets version.

Reviewer: > the only supported animated format

This is false. You can add videos as webm and they will be autoconverted to mp4 and run fine, as in #692. This should be our preferred format. Linking to movies on GH is what doesn't work.

Upsides of gif:

- can just link to a gif hosted on GH's issues

Downsides of gif:

- huge bandwidth hog — pages take much longer to load for the same video duration
- even though it has lower framerate than a comparable (and much smaller) mp4/webm
- GH could remove access to external links to user files at any moment (this is the main reason we have two repos).

Reviewer 2: So we should encourage to host the videos/webm somewhere and then use those?

Reviewer: they are hosted in this repo under docs/_static/images (from memory)

Reviewer: and to be clear, they *should* be hosted in this repo because that's how the automagic webm->mp4 conversion happens. If you want to store them elsewhere, that's fine, but you need to do your own conversion.

Reviewer 2: "they" meaning webm then, not mpf4, correct?

Reviewer: Yes, webm's (and fallback jpgs/pngs) should be hosted on this repo.

Author: but don't webm/mp4 not show in github version's of release notes? This is becasue it can't see what we have in the docs repo. I could be wrong, and would be happy for you to update the text to fix whatever I still have wrong.

Reviewer: I kinda don't care at all about the github release notes? They are there for convenience but there should be one place where people check release notes and that's the docs.

Reviewer: And I certainly don't think we should sacrifice our release notes' quality to have them show up on GH.

idea though: could we have the fallback image be the GH gif?

Author: @jni totally hear you. I think this might be a slightly generational thing? Most newer folks (myself included) to OSS/Github, in my experience, check Github release notes first (and perhaps only). It might be because the tooling for that has greatly improved over the years -- perhaps the easy autogeneration of release notes with the Github website is a promoter of that.

I totally understand that webm/mp4 would be preferred. I think, if we do try GIFs as fallbacks, we'd want to very clearly write up this pipeline for release managers. One other place we stick release notes is on image.sc (and sometimes on socials); I'd think having hardcoded links there would also be easier for sharing in those places.

Reviewer: Another idea: I wonder whether adding the GH video *at the top* of the video tag cascade would work everywhere...
